cirq-core/cirq/transformers/transformer_api.py

```diff
@@ -0,0 +1,327 @@
+# limitations under the License.
+
+"""Defines the API for circuit transformers in Cirq."""
+
+import textwrap
+import functools
+from typing import (
+    Callable,
+    Tuple,
+    TypeVar,
+    Hashable,
+    List,
+    cast,
+    Type,
+    overload,
+    TYPE_CHECKING,
+)
+import dataclasses
+import enum
+
+from cirq.circuits.circuit import CIRCUIT_TYPE
+
+if TYPE_CHECKING:
+    import cirq
+
+
+class LogLevel(enum.Enum):
+    """Different logging resolution options for `TransformerStatsLogger`.
+
+    Args:
+        ALL:     All levels. Used to filter logs when printing.
+        DEBUG:   Designates fine-grained informational events that are most useful to debug /
+                 understand in-depth any unexpected behavior of the transformer.
+        INFO:    Designates informational messages that highlight the actions of a transformer.
+        WARNING: Designates unwanted or potentially harmful situations.
+        NONE:    No levels. Used to filter logs when printing.
+    """
+
+    ALL = 0
+    DEBUG = 10
+    INFO = 20
+    WARNING = 30
+    NONE = 40
+
+
+@dataclasses.dataclass
+class LoggerNode:
+    transformer_id: int
+    transformer_name: str
+    initial_circuit: 'cirq.AbstractCircuit'
+    final_circuit: 'cirq.AbstractCircuit'
+    logs: Tuple[Tuple[LogLevel, Tuple[str, ...]], ...] = ()
+    nested_loggers: Tuple[int, ...] = ()
+
+
+class TransformerStatsLogger:
+    """Abstract Base Class for transformer logging infrastructure.
+
+    The logger implementation should be stateful, s.t.:
+        - Each call to `register_initial` registers a new transformer stage and initial circuit.
+        - Each subsequent call to `log` should store additional logs corresponding to the stage.
+        - Each call to `register_final` should register the end of the currently active stage.
+
+    The logger assumes that
+        - Transformers are run sequentially.
+        - Nested transformers are allowed, in which case the behavior would be similar to a
+          doing a depth first search on the graph of transformers -- i.e. the top level transformer
+          would end (i.e. receive a `register_final` call) once all nested transformers (i.e. all
+          `register_initial` calls received while the top level transformer was active) have
+          finished (i.e. corresponding `register_final` calls have also been received).
+        - This behavior can be simulated by maintaining a stack of currently active stages and
+          adding data from `log` calls to the stage at the top of the stack.
+
+    The `LogLevel`s can be used to control the input processing and output resolution of the logs.
+    """
+
+    def __init__(self):
+        """Initializes TransformerStatsLogger."""
+        self._curr_id: int = 0
+        self._logs: List[LoggerNode] = []
+        self._stack: List[int] = []
+
+    def register_initial(self, circuit: 'cirq.AbstractCircuit', transformer_name: str) -> None:
```

Comment:
Should the circuit type here be abstractcircuit or CIRCUIT_TYPE to match register_final ?

Reply:
Since the logger will be shared across all transformers, we should use `cirq.AbstractCircuit` instead of `CIRCUIT_TYPE`. Updated `register_final` to match this.